src/nn/transformer.rs

```diff
@@ -0,0 +1,1097 @@
+        self.w_o.forward(tokens)
+    }
+}
+
+/// A single transformer block containing self attention, feed forward and layer norms
+#[derive(Debug, Clone, Default)]
+pub struct TransformerBlock<const M: usize, const I: usize, const K: usize, const H: usize> {
+    attn: MultiHeadAttention<M, M, K, M, H>,
+    norm1: LayerNorm1D<M>,
+    norm2: LayerNorm1D<M>,
+    ff: (Linear<M, I>, ReLU, Linear<I, M>),
+}
+
+impl<const M: usize, const I: usize, const K: usize, const H: usize> ResetParams
+    for TransformerBlock<M, I, K, H>
+{
+    fn reset_params<R: Rng>(&mut self, rng: &mut R) {
+        self.attn.reset_params(rng);
+        self.norm1.reset_params(rng);
+        self.norm2.reset_params(rng);
+        self.ff.reset_params(rng);
+    }
+}
+
+impl<const M: usize, const I: usize, const K: usize, const H: usize> CanUpdateWithGradients
+    for TransformerBlock<M, I, K, H>
+{
+    fn update<G: GradientProvider>(&mut self, grads: &mut G) {
+        self.attn.update(grads);
+        self.norm1.update(grads);
+        self.norm2.update(grads);
+        self.ff.update(grads);
+    }
+}
+
+/// Single sequence impl
+impl<const M: usize, const I: usize, const K: usize, const S: usize, const H: usize>
+    Module<Tensor2D<S, M>> for TransformerBlock<M, I, K, H>
+where
+    Assert<{ H * S * (M / H) == S * M }>: ConstTrue,
+    Assert<{ S * K == H * S * (K / H) }>: ConstTrue,
+    Assert<{ S * M == H * S * (M / H) }>: ConstTrue,
+{
+    type Output = Tensor2D<S, M>;
+
+    fn forward(&self, input: Tensor2D<S, M>) -> Self::Output {
+        let x = self
+            .norm1
+            .forward(input.duplicate() + &self.attn.forward(input));
+        self.norm2.forward(x.duplicate() + &self.ff.forward(x))
+    }
+}
+
+/// Batch sequence impl
+impl<
+        const M: usize,
+        const I: usize,
+        const K: usize,
+        const S: usize,
+        const B: usize,
+        const H: usize,
+    > Module<Tensor3D<B, S, M>> for TransformerBlock<M, I, K, H>
+where
+    Assert<{ B * H * S * (M / H) == B * S * M }>: ConstTrue,
+    Assert<{ B * S * K == B * H * S * (K / H) }>: ConstTrue,
+    Assert<{ B * S * M == B * H * S * (M / H) }>: ConstTrue,
+{
+    type Output = Tensor3D<B, S, M>;
+
+    fn forward(&self, input: Tensor3D<B, S, M>) -> Self::Output {
+        let x = self
+            .norm1
+            .forward(input.duplicate() + &self.attn.forward(input));
+        self.norm2.forward(x.duplicate() + &self.ff.forward(x))
+    }
+}
+
+/// A transformer encoder.
+///
+/// # Generics
+/// - `M` The embedding size of token vectors.
+/// - `I` The inner size of the feedforward layers.
+/// - `L` The number of layers.
+/// - `H` The number of heads for self attention.
+/// TODO: Doctests
+#[derive(Debug, Clone)]
+pub struct TransformerEncoder<const M: usize, const I: usize, const L: usize, const H: usize>
+where
+    Assert<{ M % H == 0 }>: ConstTrue,
+{
+    blocks: Repeated<TransformerBlock<M, I, M, H>, L>,
+}
```

Comment:
Similarly I think this could just be
```rust
type TransformerEncoder<...> = Repeated<TransformerBlock<...>, L>;
```

Reply:
I would want to do that, but notice the where statement. Technically this could be gone, but I think it's nice that it checks that the embedding dims are dividable by the num of heads when the model is created, rather than when forward is called.

Reply:
I see what you're saying, but I don't think that alone is worth all the extra code & tests that come with it.